File: Algorithms/Easy/125.Valid-Palindrome/ValidPalindrome.py
# Time complexity: O(n)
# Space complexity: O(n)
class Solution:
    def isPalindrome(self, s: str) -> bool:
        buf = []
        for c in s.lower():
            if c.isdigit() or c.isalpha():
                buf.append(c)

        s = "".join(buf)
        l, r = 0, len(s) - 1
        while l <= r:
            if s[l] != s[r]:
                return False
            
            l += 1
            r -= 1
        
        return True

File: Algorithms/Easy/125.Valid-Palindrome/ValidPalindrome.go
package easy

import "strings"

// Time complexity: O(n)
// Space complexity: O(1)
func isPalindrome(s string) bool {
	l, r := 0, len(s)-1
	s = strings.ToLower(s)

	for l < r {
		if !isAlphaNum(s[l]) {
			l++
		} else if !isAlphaNum(s[r]) {
			r--
		} else {
			if s[l] != s[r] {
				return false
			}

			l++
			r--
		}
	}

	return true
}

func isAlphaNum(c byte) bool {
	if c >= 48 && c <= 57 || c >= 97 && c <= 122 {
		return true
	}

	return false
}
